File: cpp/digitama/basis/drunkard.cpp
#include "drunkard.hpp"

#include "../big_bang/physics/random.hpp"

using namespace WarGrey::STEM;

static const float step_size = 2.0F;
static const double step_duration = 0.2;

/*************************************************************************************************/
void WarGrey::STEM::DrunkardWalkWorld::load(float width, float height) {
    this->beach = this->insert(new Sprite(digimon_path("asset/beach", ".png")));
    this->tent = this->insert(new SpriteGridSheet(digimon_path("asset/tents", ".png"), 1, 4));
    this->drunkard = this->insert(new Agate());
    this->partner = this->insert(new Tita());

    TheBigBang::load(width, height);
}

void WarGrey::STEM::DrunkardWalkWorld::reflow(float width, float height) {
    this->move_to(this->beach, width * 0.5F, height, MatterAnchor::CB);
    this->move_to(this->tent, 0.0F, height, MatterAnchor::LB);
    
    TheBigBang::reflow(width, height);
}

void WarGrey::STEM::DrunkardWalkWorld::on_mission_start(float width, float height) {
    this->drunkard->switch_mode(BracerMode::Walk);
    this->drunkard->set_heading(-180.0);

    this->move_to(this->drunkard, width * 0.95F, height * 0.9F, MatterAnchor::CC);
    this->move_to(this->partner, width * 0.24F, height * 0.9F, MatterAnchor::CC);
}

void WarGrey::STEM::DrunkardWalkWorld::update(uint64_t interval, uint32_t count, uint64_t uptime) {
    if (!this->is_colliding(this->drunkard, this->partner)) {
        if (this->partner->motion_stopped()) {
            this->sibling_walk();
        }

        this->drunkard_walk();
    } else if (this->partner->current_mode() != BracerMode::Win) {
        this->partner->motion_stop();
        this->drunkard->switch_mode(BracerMode::Win, 1);
        this->partner->switch_mode(BracerMode::Win, 1);
    }
}

/*************************************************************************************************/
void WarGrey::STEM::DrunkardWalkWorld::sibling_walk() {
    // random_uniform(-1, 1) 长生一个位于区间 [-1, 1] 的随机数
    int dx = (random_uniform(-1, 1)); // 左右移动或不动
    int dy = (random_uniform(-1, 1)); // 上下移动或不动

    this->glide(step_duration, this->partner, dx * step_size, dy * step_size);
}

void WarGrey::STEM::DrunkardWalkWorld::drunkard_walk() {
    // 产生位于区间 [1, 100] 的随机数
    int chance = random_uniform(1, 100);
    double dx = 0.0;
    double dy = 0.0;
    
    if (chance < 10) {
        // no move
    } else if (chance < 58) {
        dx = -1.0;
    } else if (chance < 60) {
        dx = +1.0;
    } else if (chance < 80) {
        dy = +1.0;
    } else {
        dy = -1.0;
    }

    this->move(this->drunkard, dx, dy);
}

File: python/Drunkard.py
#!/usr/bin/env python3

from digitama.big_bang.game import *    # 导入游戏模块，内含 Plane 类和常用函数

###############################################################################
# 本游戏中的常量设定
step_size = 2.0
step_duration = 0.2

###############################################################################
class DrunkardWalkWorld(Plane):
    def __init__(self):
        # 通过父类的构造函数设置窗口标题和帧频
        super(DrunkardWalkWorld, self).__init__("醉汉漫步")

        # 本游戏世界有以下物体
        self.drunkard = None
        self.partner = None
        self.beach = None
        self.tent = None

    def load(self, Width, Height):
        self.beach = self.insert(Sprite(digimon_path("asset/beach", ".png")))
        self.drunkard = self.insert(Agate())
        self.partner = self.insert(Tita())

    def reflow(self, width, height):
        self.move_to(self.beach, (width * 0.5, height), MatterAnchor.CB)

    def update(self, count, interval, uptime):
        if not self.is_colliding(self.drunkard, self.partner):
            if self.partner.motion_stopped():
                self.__sibling_walk()

            self.__drunkard_walk()
        else:
            self.partner.motion_stop()
            self.drunkard.switch_mode(BracerMode.Win, 1)
            self.partner.switch_mode(BracerMode.Win, 1)

    def can_select(self, matter):
        return True
    
# protected
    def on_mission_start(self, width, height):
        self.drunkard.switch_mode(BracerMode.Walk)
        self.drunkard.set_heading(-180.0)

        self.move_to(self.drunkard, (width * 0.95, height * 0.9), MatterAnchor.CC)
        self.move_to(self.partner, (width * 0.24, height * 0.9), MatterAnchor.LC)

# private
    def __sibling_walk(self):
        dx = random.randint(-1, 1)
        dy = random.randint(-1, 1)

        self.glide(step_duration, self.partner, dx * step_size, dy * step_size)

    def __drunkard_walk(self):
        chance = random.randint(1, 100)
        dx = 0.0
        dy = 0.0
        
        if chance < 10:
            pass # no move    
        elif chance < 58:
            dx = -1.0
        elif chance < 60:
            dx = +1.0
        elif chance < 80:
            dy = +1.0
        else:
            dy = -1.0

        self.move(self.drunkard, dx, dy)

###############################################################################
launch_universe(DrunkardWalkWorld, __name__)
